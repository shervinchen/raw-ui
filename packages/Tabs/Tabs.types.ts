import { HTMLAttributes, ReactNode } from 'react';

interface BaseTabsProps {
  value?: string;
  defaultValue?: string;
  className?: string;
  disabled?: boolean;
  vertical?: boolean;
  onChange?: (value: string) => void;
}

interface BaseTabsItemProps {
  label: ReactNode;
  value: string;
  disabled?: boolean;
  className?: string;
}

type NativeTabsProps = Omit<
  HTMLAttributes<HTMLDivElement>,
  keyof BaseTabsProps
>;

type NativeTabsItemProps = Omit<
  HTMLAttributes<HTMLDivElement>,
  keyof BaseTabsItemProps
>;

export type TabsProps = BaseTabsProps & NativeTabsProps;

export type TabsItemProps = BaseTabsItemProps & NativeTabsItemProps;

export interface TabsConfig {
  selectValue?: string;
}
