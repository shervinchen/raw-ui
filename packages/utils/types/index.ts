export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Record<string, any> ? DeepPartial<T[P]> : T[P];
};

export type Merge<T, P> = P & Omit<T, keyof P>;
