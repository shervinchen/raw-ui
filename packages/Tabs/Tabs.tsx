import React, {
  FC,
  PropsWithChildren,
  useCallback,
  useMemo,
  useState,
} from 'react';
import classNames from 'classnames';
import { TabsConfig, TabsItemProps, TabsProps } from './Tabs.types';
import { useControlled } from '../utils/hooks';
import { getValidChildren } from '../utils/common';
import { TabsContext } from './tabs-context';
import { useTheme } from '../Theme';

const Tabs: FC<PropsWithChildren<TabsProps>> = ({
  defaultValue,
  value,
  className = '',
  disabled: tabsDisabled = false,
  vertical = false,
  onChange,
  children,
  ...restProps
}) => {
  const [internalValue, setInternalValue] = useControlled<string>({
    defaultValue,
    value,
  });
  const theme = useTheme();
  const classes = classNames(
    'raw-tabs',
    vertical && 'raw-tabs-vertical',
    className
  );
  const [navBarPosition, setNavBarPosition] = useState({
    width: 0,
    height: 0,
    left: 0,
    top: 0,
  });

  const tabsNav = useMemo(() => {
    return getValidChildren(children).map((child) => {
      const { label, value, disabled = false } = child.props as TabsItemProps;
      return {
        label,
        value,
        disabled,
      };
    });
  }, [children]);

  const tabsConfig: TabsConfig = {
    selectValue: internalValue,
  };

  const tabsNavRef = useCallback(
    (node: HTMLDivElement) => {
      const navItems = [...(node?.children ?? [])] as HTMLDivElement[];
      const selectedNavIndex = tabsNav.findIndex(
        (item) => item.value === internalValue
      );
      const selectedNavItem = navItems
        .filter((item) => item.role === 'tab')
        .find((_, index) => index === selectedNavIndex);
      if (selectedNavItem) {
        const { offsetLeft, offsetTop } = selectedNavItem;
        const { width, height } = selectedNavItem.getBoundingClientRect();
        setNavBarPosition({
          width,
          height,
          left: offsetLeft,
          top: offsetTop,
        });
      }
    },
    [internalValue, tabsNav]
  );

  const handleTabsNavItemClick = (value: string, disabled: boolean) => {
    const isDisabled = tabsDisabled || disabled;
    if (isDisabled) return;
    setInternalValue(value);
    onChange?.(value);
  };

  return (
    <TabsContext.Provider value={tabsConfig}>
      <div className={classes} {...restProps}>
        <div role="tablist" ref={tabsNavRef} className="raw-tabs-nav">
          {tabsNav.map((item) => (
            <div
              role="tab"
              className={classNames(
                'raw-tabs-nav-item',
                internalValue === item.value && 'raw-tabs-nav-item-active',
                (tabsDisabled || item.disabled) && 'raw-tabs-nav-item-disabled'
              )}
              key={item.value}
              onClick={() => handleTabsNavItemClick(item.value, item.disabled)}
            >
              {item.label}
            </div>
          ))}
          <div
            className="raw-tabs-nav-bar"
            style={{
              width: vertical ? '2px' : navBarPosition.width,
              height: vertical ? navBarPosition.height : '2px',
              transform: vertical
                ? `translateY(${navBarPosition.top}px)`
                : `translateX(${navBarPosition.left}px)`,
            }}
          />
        </div>
        <div className="raw-tabs-content">{children}</div>
        <style jsx>
          {`
            .raw-tabs {
              display: flex;
              flex-direction: column;
              width: 100%;
              gap: 16px;
            }
            .raw-tabs.raw-tabs-vertical {
              flex-direction: row;
            }
            .raw-tabs .raw-tabs-nav {
              position: relative;
              display: flex;
              align-items: center;
              flex-wrap: nowrap;
              flex-direction: row;
              gap: 24px;
              overflow-x: auto;
              overflow-y: hidden;
              scrollbar-width: none;
            }
            .raw-tabs .raw-tabs-nav::before {
              position: absolute;
              top: ${navBarPosition.height - 1}px;
              left: 0;
              content: '';
              width: 100%;
              height: 1px;
              border-bottom: 1px solid ${theme.palette.accents2};
            }
            .raw-tabs.raw-tabs-vertical .raw-tabs-nav {
              flex-direction: column;
              gap: 12px;
              overflow-x: hidden;
              overflow-y: auto;
            }
            .raw-tabs.raw-tabs-vertical .raw-tabs-nav::before {
              top: 0;
              left: calc(100% - 1px);
              width: 1px;
              height: 100%;
              border-bottom: none;
              border-right: 1px solid ${theme.palette.accents2};
            }
            .raw-tabs .raw-tabs-nav-item {
              padding: 12px 0;
              font-size: 16px;
              color: ${theme.palette.accents7};
              background-color: transparent;
              white-space: nowrap;
              user-select: none;
              cursor: pointer;
              transition-property: color;
              transition-duration: 0.2s;
              transition-timing-function: ease;
            }
            .raw-tabs.raw-tabs-vertical .raw-tabs-nav-item {
              padding: 6px 12px;
            }
            .raw-tabs-nav-item.raw-tabs-nav-item-active {
              color: ${theme.palette.foreground};
            }
            .raw-tabs-nav-item.raw-tabs-nav-item-disabled {
              color: ${theme.palette.accents5};
              cursor: not-allowed;
            }
            .raw-tabs-nav-item:not(.raw-tabs-nav-item-active):not(
                .raw-tabs-nav-item-disabled
              ):hover {
              color: ${theme.palette.foreground};
            }
            .raw-tabs .raw-tabs-nav-bar {
              position: absolute;
              top: ${navBarPosition.height - 2}px;
              left: 0;
              background-color: ${theme.palette.foreground};
              transition-property: transform;
              transition-duration: 0.2s;
              transition-timing-function: ease;
            }
            .raw-tabs.raw-tabs-vertical .raw-tabs-nav-bar {
              top: 0;
              left: calc(100% - 2px);
            }
          `}
        </style>
      </div>
    </TabsContext.Provider>
  );
};

export default Tabs;
