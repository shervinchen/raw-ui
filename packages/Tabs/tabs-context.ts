import { createContext, useContext } from 'react';
import { TabsConfig } from './Tabs.types';

const defaultContext: TabsConfig = {
  tabsId: '',
};

export const TabsContext = createContext<TabsConfig>(defaultContext);

export const useTabsContext = (): TabsConfig =>
  useContext<TabsConfig>(TabsContext);
