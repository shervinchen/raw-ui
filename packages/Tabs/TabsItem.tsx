import React, { FC, PropsWithChildren } from 'react';
import classNames from 'classnames';
import { TabsItemProps } from './Tabs.types';
import { useTabsContext } from './tabs-context';

const TabsItem: FC<PropsWithChildren<TabsItemProps>> = ({
  value,
  className = '',
  children,
}) => {
  const { tabsId, selectValue } = useTabsContext();
  const classes = classNames(
    'raw-tabs-panel',
    selectValue === value && 'raw-tabs-panel-active',
    className
  );

  return (
    <div
      role="tabpanel"
      id={`raw-tabs-panel-${value}-${tabsId}`}
      aria-labelledby={`raw-tabs-nav-item-${value}-${tabsId}`}
      aria-hidden={selectValue !== value}
      className={classes}
    >
      {children}
      <style jsx>
        {`
          .raw-tabs-panel {
            display: none;
          }
          .raw-tabs-panel.raw-tabs-panel-active {
            display: block;
          }
        `}
      </style>
    </div>
  );
};

export default TabsItem;
