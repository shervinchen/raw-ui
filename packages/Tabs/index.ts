import Tabs from './Tabs';
import TabsItem from './TabsItem';

export type TabsComponentType = typeof Tabs & {
  Tab: typeof TabsItem;
};

(Tabs as TabsComponentType).Tab = TabsItem;

export default Tabs as TabsComponentType;
