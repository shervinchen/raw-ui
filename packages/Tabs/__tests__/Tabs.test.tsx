import React, { useState } from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Tabs } from '../..';
import { TabsProps } from '../Tabs.types';

const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

describe('Tabs', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(
      <Tabs defaultValue="1">
        <Tabs.Tab label="Tab 1" value="1">
          Content of Tab Pane 1
        </Tabs.Tab>
        <Tabs.Tab label="Tab 2" value="2">
          Content of Tab Pane 2
        </Tabs.Tab>
        <Tabs.Tab label="Tab 3" value="3">
          Content of Tab Pane 3
        </Tabs.Tab>
      </Tabs>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should support custom class name', () => {
    render(
      <Tabs defaultValue="1" className="custom-tabs">
        <Tabs.Tab label="Tab 1" value="1">
          Content of Tab Pane 1
        </Tabs.Tab>
        <Tabs.Tab label="Tab 2" value="2">
          Content of Tab Pane 2
        </Tabs.Tab>
        <Tabs.Tab label="Tab 3" value="3">
          Content of Tab Pane 3
        </Tabs.Tab>
      </Tabs>
    );
    expect(screen.getByTestId('tabs')).toHaveClass('custom-tabs');
  });

  test('should support default value', () => {
    render(
      <Tabs defaultValue="2">
        <Tabs.Tab label="Tab 1" value="1">
          Content of Tab Pane 1
        </Tabs.Tab>
        <Tabs.Tab label="Tab 2" value="2">
          Content of Tab Pane 2
        </Tabs.Tab>
        <Tabs.Tab label="Tab 3" value="3">
          Content of Tab Pane 3
        </Tabs.Tab>
      </Tabs>
    );
    expect(screen.getByText('Tab 2')).toHaveClass('raw-tabs-nav-item-active');
  });

  test('should support controlled value', async () => {
    const onChange = jest.fn();

    const Component = (props: TabsProps) => {
      const [value, setValue] = useState('1');

      return (
        <Tabs
          value={value}
          onChange={(newValue) => {
            setValue(newValue);
            props.onChange?.(newValue);
          }}
        >
          <Tabs.Tab label="Tab 1" value="1">
            Content of Tab Pane 1
          </Tabs.Tab>
          <Tabs.Tab label="Tab 2" value="2">
            Content of Tab Pane 2
          </Tabs.Tab>
          <Tabs.Tab label="Tab 3" value="3">
            Content of Tab Pane 3
          </Tabs.Tab>
        </Tabs>
      );
    };

    render(<Component onChange={onChange} />);
    const tabItemSecond = screen.getByText('Tab 2');
    await user.click(tabItemSecond);
    expect(tabItemSecond).toHaveClass('raw-tabs-nav-item-active');
    expect(onChange).toHaveBeenCalledWith('2');
  });

  test('should not select any tab when no value matched', () => {
    render(
      <Tabs defaultValue="4">
        <Tabs.Tab label="Tab 1" value="1">
          Content of Tab Pane 1
        </Tabs.Tab>
        <Tabs.Tab label="Tab 2" value="2">
          Content of Tab Pane 2
        </Tabs.Tab>
        <Tabs.Tab label="Tab 3" value="3">
          Content of Tab Pane 3
        </Tabs.Tab>
      </Tabs>
    );
    expect(screen.getByText('Tab 1')).not.toHaveClass(
      'raw-tabs-nav-item-active'
    );
    expect(screen.getByText('Tab 2')).not.toHaveClass(
      'raw-tabs-nav-item-active'
    );
    expect(screen.getByText('Tab 3')).not.toHaveClass(
      'raw-tabs-nav-item-active'
    );
  });

  test('should not render any tab when child tabs is empty', () => {
    render(<Tabs defaultValue="1"></Tabs>);
    expect(screen.queryByText('Tab 1')).not.toBeInTheDocument();
    expect(screen.queryByText('Tab 2')).not.toBeInTheDocument();
    expect(screen.queryByText('Tab 3')).not.toBeInTheDocument();
  });

  test('should support all tabs disabled', async () => {
    render(
      <Tabs defaultValue="1" disabled>
        <Tabs.Tab label="Tab 1" value="1">
          Content of Tab Pane 1
        </Tabs.Tab>
        <Tabs.Tab label="Tab 2" value="2">
          Content of Tab Pane 2
        </Tabs.Tab>
        <Tabs.Tab label="Tab 3" value="3">
          Content of Tab Pane 3
        </Tabs.Tab>
      </Tabs>
    );
    expect(screen.getByText('Tab 1')).toHaveClass('raw-tabs-nav-item-disabled');
    expect(screen.getByText('Tab 2')).toHaveClass('raw-tabs-nav-item-disabled');
    expect(screen.getByText('Tab 3')).toHaveClass('raw-tabs-nav-item-disabled');
    await user.click(screen.getByText('Tab 2'));
    expect(screen.getByText('Tab 2')).not.toHaveClass(
      'raw-tabs-nav-item-active'
    );
  });

  test('should support tab item disabled', async () => {
    render(
      <Tabs defaultValue="1">
        <Tabs.Tab label="Tab 1" value="1">
          Content of Tab Pane 1
        </Tabs.Tab>
        <Tabs.Tab label="Tab 2" value="2" disabled>
          Content of Tab Pane 2
        </Tabs.Tab>
        <Tabs.Tab label="Tab 3" value="3">
          Content of Tab Pane 3
        </Tabs.Tab>
      </Tabs>
    );
    expect(screen.getByText('Tab 2')).toHaveClass('raw-tabs-nav-item-disabled');
    await user.click(screen.getByText('Tab 2'));
    expect(screen.getByText('Tab 2')).not.toHaveClass(
      'raw-tabs-nav-item-active'
    );
  });

  test('should support vertical', () => {
    render(
      <Tabs defaultValue="1" vertical>
        <Tabs.Tab label="Tab 1" value="1">
          Content of Tab Pane 1
        </Tabs.Tab>
        <Tabs.Tab label="Tab 2" value="2">
          Content of Tab Pane 2
        </Tabs.Tab>
        <Tabs.Tab label="Tab 3" value="3">
          Content of Tab Pane 3
        </Tabs.Tab>
      </Tabs>
    );

    expect(screen.getByTestId('tabs')).toHaveClass('raw-tabs-vertical');
  });
});
