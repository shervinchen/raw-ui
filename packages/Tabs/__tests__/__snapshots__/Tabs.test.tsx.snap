// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-2406371952 raw-tabs"
    data-testid="tabs"
  >
    <div
      aria-orientation="horizontal"
      class="jsx-2406371952 raw-tabs-nav"
      role="tablist"
    >
      <div
        aria-controls="raw-tabs-panel-1-:r0:"
        aria-disabled="false"
        aria-selected="true"
        class="jsx-2406371952 raw-tabs-nav-item raw-tabs-nav-item-active"
        id="raw-tabs-nav-item-1-:r0:"
        role="tab"
      >
        Tab 1
      </div>
      <div
        aria-controls="raw-tabs-panel-2-:r0:"
        aria-disabled="false"
        aria-selected="false"
        class="jsx-2406371952 raw-tabs-nav-item"
        id="raw-tabs-nav-item-2-:r0:"
        role="tab"
      >
        Tab 2
      </div>
      <div
        aria-controls="raw-tabs-panel-3-:r0:"
        aria-disabled="false"
        aria-selected="false"
        class="jsx-2406371952 raw-tabs-nav-item"
        id="raw-tabs-nav-item-3-:r0:"
        role="tab"
      >
        Tab 3
      </div>
      <div
        class="jsx-2406371952 raw-tabs-nav-bar"
        style="width: 0px; height: 2px; transform: translateX(0px);"
      />
    </div>
    <div
      class="jsx-2406371952 raw-tabs-content"
    >
      <div
        aria-hidden="false"
        aria-labelledby="raw-tabs-nav-item-1-:r0:"
        class="jsx-286161192 raw-tabs-panel raw-tabs-panel-active"
        id="raw-tabs-panel-1-:r0:"
        role="tabpanel"
      >
        Content of Tab Pane 1
      </div>
      <div
        aria-hidden="true"
        aria-labelledby="raw-tabs-nav-item-2-:r0:"
        class="jsx-286161192 raw-tabs-panel"
        id="raw-tabs-panel-2-:r0:"
        role="tabpanel"
      >
        Content of Tab Pane 2
      </div>
      <div
        aria-hidden="true"
        aria-labelledby="raw-tabs-nav-item-3-:r0:"
        class="jsx-286161192 raw-tabs-panel"
        id="raw-tabs-panel-3-:r0:"
        role="tabpanel"
      >
        Content of Tab Pane 3
      </div>
    </div>
  </div>
</DocumentFragment>
`;
